resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -200,58 +224,94 @@ private[spark] class KubernetesClusterSchedulerBackend(
+            .addToLimits("memory", executorMemoryLimitQuantity)
+            .addToRequests("cpu", executorCpuQuantity)
+            .addToLimits("cpu", executorCpuQuantity)
+          .endResources()
+          .withEnv(requiredEnv.asJava)
+          .withPorts(requiredPorts.asJava)
+        .endContainer()
+      .endSpec()
+      .build()
+
+      shuffleDir match {
+        case Some(dir) =>
+          pod.getSpec().setVolumes(List(new VolumeBuilder()
+            .withName(DEFAULT_SHUFFLE_MOUNT_NAME)
+            .withNewHostPath().withPath(dir)
+            .endHostPath()
+            .build()).asJava)
+
+          pod.getSpec().getContainers().get(0)
+            .setVolumeMounts(List(new VolumeMountBuilder()
+              .withName(DEFAULT_SHUFFLE_MOUNT_NAME)
+              .withMountPath(dir).build()).asJava)
+
+        case None =>
+      }
+
+    try {
+      (executorId, kubernetesClient.pods().create(pod))
     } catch {
       case throwable: Throwable =>
         logError("Failed to allocate executor pod.", throwable)
         throw throwable
     }
   }
 
-  override def doRequestTotalExecutors(requestedTotal: Int): Future[Boolean] = Future[Boolean] {
-    EXECUTOR_MODIFICATION_LOCK.synchronized {
-      if (requestedTotal > totalExpectedExecutors.get) {
-        logInfo(s"Requesting ${requestedTotal - totalExpectedExecutors.get}"
-          + s" additional executors, expecting total $requestedTotal and currently" +
-          s" expected ${totalExpectedExecutors.get}")
-        for (i <- 0 until (requestedTotal - totalExpectedExecutors.get)) {
-          runningExecutorPods += allocateNewExecutorPod()
+  override def doRequestTotalExecutors(requestedTotal: Int): Future[Boolean] =
+    Future[Boolean] {
+      EXECUTOR_MODIFICATION_LOCK.synchronized {
+        if (requestedTotal > totalExpectedExecutors.get) {
+          logInfo(s"Requesting ${requestedTotal - totalExpectedExecutors.get}"
+            + s" additional executors, expecting total $requestedTotal and currently" +
+            s" expected ${totalExpectedExecutors.get}")
+          for (i <- 0 until (requestedTotal - totalExpectedExecutors.get)) {
+            runningExecutorPods += allocateNewExecutorPod()
+          }
         }
+        totalExpectedExecutors.set(requestedTotal)
       }
-      totalExpectedExecutors.set(requestedTotal)
+
+      var allPodsReady = true
+      for (executorPod <- runningExecutorPods.values) {
+        try {
+          val pod = kubernetesClient.pods().inNamespace(kubernetesNamespace).
+            withName(executorPod.getMetadata().getName()).get()
+
+          allPodsReady &&= pod.getStatus.getConditions.asScala
+            .toList.exists(cond => cond.getType == "Ready" && cond.getStatus == "True")
+        } catch {
+          case throwable: Throwable =>
+            logWarning(s"Executor no longer exists.", throwable)
+            allPodsReady = false
+        }
+      }
```

Comment:
My understanding of the API is that this is supposed to have "fire and forget" semantics, where the executors are requested but there's no guarantee that they are up after they have been requested. The return value of the future is only if the request is acknowledged, not if the executors resolved in starting up.